lib/pkce.js

```diff
@@ -0,0 +1,109 @@
+/*!
+ * Copyright (c) 2015-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ *
+ */
+
+ /* eslint-disable complexity, max-statements */
+var http          = require('./http');
+var util          = require('./util');
+var oauthUtil     = require('./oauthUtil');
+var AuthSdkError  = require('./errors/AuthSdkError');
+var token      = require('./token');
+
+// Code verifier: Random URL-safe string with a minimum length of 43 characters.
+// Code challenge: Base64 URL-encoded SHA-256 hash of the code verifier.
+var MIN_VERIFIER_LENGTH = 43;
+function generateVerifier(prefix) {
+  prefix = prefix || '';
+  var str = prefix;
+  if (str.length < MIN_VERIFIER_LENGTH) {
```

Comment:
It is important to validate that the verifier [does not exceed a maximum length of 128 chars](https://tools.ietf.org/html/rfc7636#section-4.1).

Right now, if someone passes in a `prefix` with length > 43, that value will be used as the verifier, which may **not** adhere to the specification's guidelines.

Reply:
Added functionality to trim the string to maximum length, along with a test